### ISSUE TITLE
a nan error !

### ISSUE THREAD

**Issue by qianqing13579** [ROCmSoftwarePlatform/AMDMIGraphX#831]
I use migraphx to do inference,but the output is nan. When I use onnx python to do the same thing,the result is OK. Can you give me some advice?
the model link: https://pan.baidu.com/s/1xQlpe6zXysx9BB-JyCqPuw 
password：vj9b

**Comment by scxiao** [comment]
I tried it, looks like it works fine. 
What is the program that you use to run the model?

**Comment by qianqing13579** [comment]
@scxiao ,the code:
`#include <string>
#include <vector>
#include <migraphx/onnx.hpp>
#include <migraphx/gpu/target.hpp>
#include <migraphx/gpu/hip.hpp>
#include <migraphx/generate.hpp>
#include <migraphx/quantization.hpp>


using namespace std;
using namespace migraphx;

int main(int argc, char *argv[])
{
    // load model
    migraphx::program net= migraphx::parse_onnx("/LocalData/Resource/Models/EfficientNet_model/model_convert_simil_NCHW_bs1.onnx");

    // use fp16
    migraphx::quantize_fp16(net);

    net.compile(migraphx::gpu::target{});

    // input
    migraphx::parameter_map m;
    for (auto&& x : net.get_parameter_shapes())
    {
        m[x.first] = migraphx::gpu::to_gpu(migraphx::generate_argument(x.second));
    }
    auto s = migraphx::shape{migraphx::shape::float_type, {1, 3, 300, 300}};
    m["input_1:0"]= migraphx::gpu::to_gpu(generate_argument(s,0)); // 将数据从CPU转换到GPU

    // infere
    auto results = net.eval(m).back();

    // get result
    auto result  = migraphx::gpu::from_gpu(results); // GPU->CPU
    std::vector<float> logits;
    result.visit([&](auto output) { logits.assign(output.begin(), output.end()); });
    for(int i=0;i<logits.size();++i)
    {
        printf("%d:%f\n",i,logits[i]);
    }

    return 0;

}`

**Comment by qianqing13579** [comment]
@scxiao ,if you use the code,the output is nan!

**Comment by scxiao** [comment]
Yes, by converting the model to fp16, I can reproduce. 
Then what do you mean for onnx python, does it use migraphx?

**Comment by qianqing13579** [comment]
If I use TensorRT,the result is right.I do not know why.
I use MIGRAPHX_TRACE_EVAL to check the output of each instruction,and I find that several outputs are nan! I think the result of some operator is wrong,may be,I am trying to find out the reason

**Comment by qianqing13579** [comment]
@scxiao ，I find some wrong results of the output.
Run instruction: @53 = gpu::convolution[padding={0, 0},stride={1, 1},dilation={1, 1},group=1,padding_mode=0,solution_id=13](@52,@49,@51,@50) -> half_type, {1, 10, 1, 1}, {10, 1, 1, 1}
Ouput: -nan, -nan, -nan, -inf, inf, inf, -inf, -nan, inf, -nan

**Comment by scxiao** [comment]
The root cause is we have an overflow in the reduce_mean() if half_type is used.
main:@425 = reduce_mean[axes={1, 2}](main:@424) -> float_type, {1, 1, 1, 40}, {40, 1, 1, 1}
Specifically, half type has max value 65504, by recording all the accumulated value in reduce mean, we observed some values like: 83316.7, 66779. These values are out of the range and we have no processing for that.

Then the output is used as input to the convolution, and the output becomes nan.

**Comment by qianqing13579** [comment]
@scxiao ，if FP32 is used,there is also an error:
Memory access fault by GPU node-2 (Agent handle: 0x146c010) on address (nil). Reason: Page not present or supervisor privilege.

**Comment by qianqing13579** [comment]
@scxiao ，if FP16 is used,how to solve the overflow in the reduce_mean() ?

**Comment by qianqing13579** [comment]
@scxiao ，I set an object to 65504 if its value is greater than 65504 in reduce_mean function,but the result is also nan!

**Comment by scxiao** [comment]
Thanks for the verification. We are working on a solution for that. Will keep you updated for that.

**Comment by qianqing13579** [comment]
@scxiao ，can you give me some advice

**Comment by scxiao** [comment]
@qianqing13579 Sorry for the late response. If you know how to build the source code, one walkaround is:
When parsing the reduce_mean operator, we can add a mul() operator like mul(x, 0.01) before the reduce_mean, then after the reduce_mean, we can add another mul() like mul(x, 100.0f) to adjust the reduce_mean output to the correct value. 

We are try to come up with a general solution to resolve this issue, but it may take some time. Thanks!

**Comment by scxiao** [comment]
@qianqing13579 For your convenience, I created a branch "fp16_reduce_mean" for the above proposed changes. You can build it for your testing. 

I tested the change, the "nan" outputs are gone.

**Comment by qianqing13579** [comment]
@scxiao ,thanks!
